### PR TITLE
[SPARK-46760][SQL][DOCS] Make the document of spark.sql.adaptive.coalescePartitions.parallelismFirst clearer

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR propose to make the document of `spark.sql.adaptive.coalescePartitions.parallelismFirst` clearer.

### Why are the changes needed?
The default value of `spark.sql.adaptive.coalescePartitions.parallelismFirst` is true, but the document contains the word `recommended to set this config to false and respect the configured target size`. It's very confused.

### Does this PR introduce _any_ user-facing change?
'Yes'.
The document is more clear.

### How was this patch tested?
N/A

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -721,8 +721,8 @@ object SQLConf {
         "shuffle partitions, but adaptively calculate the target size according to the default " +
         "parallelism of the Spark cluster. The calculated size is usually smaller than the " +
         "configured target size. This is to maximize the parallelism and avoid performance " +
-        "regression when enabling adaptive query execution. It's recommended to set this config " +
-        "to false and respect the configured target size.")
+        "regression when enabling adaptive query execution. If you respect the configured " +
+        "target size, please set this config to false.")
```
Comment (nchammas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:725]: Minor suggestion:

```suggestion
        "regressions when enabling adaptive query execution. To respect the configured " +
        "target size, please set this config to false.")
```

**Comment by nchammas** [comment]
Note that until we adopt some approach to automate the generation of config tables in our docs (e.g. like in #44755 or #44756), you will need to manually update the HTML here so it stays in sync with the source: https://github.com/apache/spark/blob/977f64f0904e46e72cbe5b2252f2657dde29c90c/docs/sql-performance-tuning.md#L270

**Review by nchammas — APPROVED** [review]

**Comment by beliefer** [comment]
cc @cloud-fan

**Comment by beliefer** [comment]
cc @MaxGekk @gengliangwang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -721,8 +721,8 @@ object SQLConf {
         "shuffle partitions, but adaptively calculate the target size according to the default " +
         "parallelism of the Spark cluster. The calculated size is usually smaller than the " +
         "configured target size. This is to maximize the parallelism and avoid performance " +
-        "regression when enabling adaptive query execution. It's recommended to set this config " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:724]: @maryannxue is it really recommended?

**Review by srowen — CHANGES_REQUESTED** [review]
Agree. I'm not sure why it says "it's recommended to be set to false" when that's not the default and 'true' is the safer default.

But this text still really doesn't help you decide anything. What does 'parallelism first' even mean? we can't change that name, but we can explain it.

Instead of all of the text starting at "This is to avoid ...", which isn't even that accurate (you can have a perf problem either way), let's continue:

"This is helpful where even small partitions with small data size require a large amount of computation, and so coalescing the small partitions reduces parallelism and harms performance. In more typical cases where this is not true, coalescing partitions can avoid many tiny tasks and improve performance, and so this config can be set to `false`"

**Comment by beliefer** [comment]
> "This is helpful where even small partitions with small data size require a large amount of computation, and so coalescing the small partitions reduces parallelism and harms performance. In more typical cases where this is not true, coalescing partitions can avoid many tiny tasks and improve performance, and so this config can be set to `false`"

This is very well.

**Comment by cloud-fan** [comment]
This is true by default because in benchmarks we run one query at a time and it can use all the resources of the entire cluster. So parallelism is more important, as if we pick 64mb as the target size and save some CPU slots, no other tasks will use these free slots.

**Comment by srowen** [comment]
(Re run the tests)
Is the text OK @cloud-fan or are you suggesting a modification?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:724]: maybe just say `It's recommended to set this config to true on a busy cluster to male resource utilization more efficient (not many small tasks).`

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:724]: OK

**Comment by srowen** [comment]
Merged to master

**Comment by beliefer** [comment]
@srowen @cloud-fan Thank you!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:724]: This suggestion contains a mistake right? It should be set to false in a busy cluster? https://github.com/apache/spark/pull/45437

**Pull request closed without merging**